test-manager/test_macro/src/lib.rs

```diff
@@ -76,52 +76,79 @@ fn create_test(
         Some(priority) => quote!{Some(#priority)},
         None => quote!{None},
     };
-    let cleanup = if test_function.macro_parameters.cleanup {
-        quote!{
-            // TODO: This hardcoded crate dependency could be avoided with a third crate for
-            // holding types such as these
-            crate::tests::cleanup_after_test(*mullvad_client).await?;
-        }
-    } else {
-        quote!{}
-    };
+    let should_cleanup = test_function.macro_parameters.cleanup;
 
     let func_name = test_function.name;
-    let wrapper_closure = if let Some(mullvad_client_type) = test_function.function_parameters.mullvad_client_type.clone() {
-        quote! {
-            |rpc: test_rpc::ServiceClient,
-            mullvad_client: Box<dyn std::any::Any + Send>,|
-            {
-                use std::any::Any;
-                let mullvad_client = mullvad_client.downcast::<#mullvad_client_type>().expect("invalid mullvad client");
-                let func = Box::pin(async move {
-                    let result = #func_name(rpc, *mullvad_client.clone()).await;
-                    #cleanup
-                    result
-                });
-                func
+    let function_mullvad_version = test_function.function_parameters.mullvad_client.version();
+    let wrapper_closure = match test_function.function_parameters.mullvad_client {
+        MullvadClient::New { mullvad_client_type, .. } => {
+            let mullvad_client_type = *mullvad_client_type;
+            quote! {
+                |rpc: test_rpc::ServiceClient,
+                mullvad_client: Box<dyn std::any::Any + Send>,|
+                {
+                    use std::any::Any;
+                    let mut mullvad_client = mullvad_client.downcast::<#mullvad_client_type>().expect("invalid mullvad client");
+                    Box::pin(async move {
+                        let result = #func_name(rpc, *mullvad_client.clone()).await;
+                        // If default settings are not retrieved, retrieve them and cleanup
+                        match crate::tests::get_default_settings() {
```

Comment:
I wonder if we could just simplify this to

```rust
let default_settings = crate::tests::get_default_settings().await.unwrap();
let mullvad_client = Some(*mullvad_client);
if #should_cleanup {
    crate::tests::cleanup_after_test(default_settings, mullvad_client).await?;
}
```

and lazily initialize the settings in `get_default_settings`.

Reply:
Fixed.